File: NodeJS/esercizio8.js
#!/usr/bin/env node
/*
Esercizio: Scrivere una classe Istogramma che permetta di aggiungere numeri, fare la stampa e poi che implementi i metodi:
double GetMean() (che calcola il valor medio)
double GetRMS() (che calcola la deviazione standard)
*/

'use strict';

function Istogramma(nBin, min, max) {
	var numeroBin = (nBin) ? nBin : 0;
	var valoreMinimo = (min) ? min : 0;
	var valoreMassimo = (max) ? max : 0;
	var binContent = new Array();
	var entries = 0;
	if (nBin) {
		for (var i = 0; i < numeroBin; ++i) {
			binContent.push(0);
		}
	}
	
	return {
		Fill: function(value) {
			if (value < valoreMinimo || value >= valoreMassimo) {
				return -1;
			}
			else {
				entries += 1;

				var invStep = numeroBin / (valoreMassimo - valoreMinimo);
				var bin = Math.floor((value - valoreMinimo) * invStep);

				binContent[bin] += 1;

				return bin;
			}
		},
		Print: function() {
			// normalizza l'istogrammma al valore maggiore
			var max = 0;
			for (var i = 0; i < numeroBin; ++i) {
				if (binContent[i] > max) {
					max = binContent[i];
				}
			}

			// fattore di dilatazione per la rappresentazione dell'istogramma
			var scale = 50;

			// disegna l'asse y
			var output = "	+---------------------------------------------------------------->\n";
			var invStep = numeroBin / (valoreMassimo - valoreMinimo);

			// disegna il contenuto dei bin
			for (var i = 0; i < numeroBin; ++i) {
				var label = (valoreMinimo + i / invStep).toFixed(2);
				output += ("        " + label).slice(8-label.length) + "|";
				var freq = Math.floor(scale * binContent[i] / max);
				for (var j = 0; j < freq; ++j) {
					output += "#";
				}

				output += "\n";
			}
			output += "	|\n";
			
			console.log(output);
		},
		GetMean: function() {
			var step = (valoreMassimo - valoreMinimo) / numeroBin;
			var sum = 0;

			for (var i = 0; i < numeroBin; ++i) {
				var val = i * step + valoreMinimo;
				sum += val * binContent[i];
			}

			return sum / entries;
		},
		GetRMS: function() {
			var step = (valoreMassimo - valoreMinimo) / numeroBin;
			var sum = 0;
			var sum2 = 0;

			for (var i = 0; i < numeroBin; ++i) {
				var val = i * step + valoreMinimo;
				sum += val * binContent[i];
				sum2 += Math.pow(val, 2) * binContent[i];
			}

			return (sum2 / entries - (sum / entries) * (sum / entries));
		},
		GetEntries: function() {
			return entries;
		}
	};
}

function RandFlat(a, b) {
	var r = a + (b - a) * Math.random();
	return r;
}

var readline = require('readline-sync');

//creo l'istogramma
console.log("Inserisci gli estremi dell'istogramma [min,max).");
var min = parseFloat(readline.question("Estremo min: "));
var max = parseFloat(readline.question("Estremo max: "));

var nBin = parseInt(readline.question("Inserisci il numero di bin dell'istogramma: "));
var numeri = parseInt(readline.question("Inserisci quanti numeri casuali vuoi generare: "));

// ctor
var histo = new Istogramma(nBin, min, max);

// riempio l'istogramma
for (var i = 0; i < numeri; ++i) {
	var number = RandFlat(min, max);
	histo.Fill(number);
}

// stampo 
console.log("Mean = " +  histo.GetMean().toFixed(2));
console.log("RMS = " + histo.GetRMS().toFixed(2));
histo.Print();

File: C++/esercizio8.cc
#include "esercizio8.h"
#include <iostream>
#include <iomanip>
#include <cmath>

// default ctor ------------------------------------------------------------------------------------
istogramma::istogramma () {
	nBin_p = 0;
	min_p = 0.;
	max_p = 0.;
	binContent_p = 0;
	entries_p = 0;
}

// ctor --------------------------------------------------------------------------------------------
istogramma::istogramma (const int &nBin, const double &min, const double &max) {
	nBin_p = nBin;
	min_p = min;
	max_p = max;
	binContent_p = new int[nBin_p];
	entries_p = 0;
	
	//azzero gli elementi dell'istogramma
	for (int i = 0; i < nBin_p; ++i) {
		binContent_p[i] = 0;
	}
}

// copy ctor ---------------------------------------------------------------------------------------
istogramma::istogramma (const istogramma& original) {
	nBin_p = original.nBin_p;
	min_p = original.min_p;
	max_p = original.max_p;
	binContent_p = new int[original.nBin_p];
	entries_p = original.entries_p;
	
	//copio gli elementi dell'istogramma
	for (int i = 0; i < nBin_p; ++i) {
		binContent_p[i] = original.binContent_p[i];
	}
}

// dtor --------------------------------------------------------------------------------------------
istogramma::~istogramma() {
	delete[] binContent_p;
}

// operator= ---------------------------------------------------------------------------------------
istogramma& istogramma::operator= (const istogramma& original) {
	if (binContent_p) delete[] binContent_p;

	nBin_p = original.nBin_p;
	min_p = original.min_p;
	max_p = original.max_p;
	binContent_p = new int[original.nBin_p];
	entries_p = original.entries_p;

	//copio gli elementi dell'istogramma
	for (int i = 0; i < nBin_p; ++i) {
		binContent_p[i] = original.binContent_p[i];
	}

	return *this;
}

// methods -----------------------------------------------------------------------------------------
int istogramma::Fill(const double& value) {
	if (value < min_p || value >= max_p) {
		return -1;
	}
	else  {
		++entries_p;
    
		double invStep = nBin_p/(max_p-min_p);
		int bin = int((value-min_p)*invStep);
		++binContent_p[bin];
		
		return bin;
	}
}

double istogramma::GetMean() const {
	double step = (max_p-min_p)/nBin_p;
	double sum = 0.;
	
	for (int i = 0; i < nBin_p; ++i) {
		double val = i * step + min_p;
		sum += val * binContent_p[i];	
	}
	
	return sum/entries_p;
}

double istogramma::GetRMS() const {
	double step = (max_p-min_p)/nBin_p;
	double sum = 0.;
	double sum2 = 0.;
	
	for (int i = 0; i < nBin_p; ++i) {
		double val = i * step + min_p;
		sum += val * binContent_p[i];
		sum2 += pow(val, 2) * binContent_p[i];
	}
	
	return (sum2/entries_p - (sum/entries_p)*(sum/entries_p));
}

void istogramma::Print() const {
	// normalizza l'istogrammma al valore maggiore
	int max = 0;
	for (int i = 0; i < nBin_p; ++i) {
		if (binContent_p[i] > max) {
			max = binContent_p[i];
		}
	}

	// fattore di dilatazione per la rappresentazione dell'istogramma
	int scale = 50;

	// disegna l'asse y
	std::cout << "        +---------------------------------------------------------------->" << std::endl;

	double invStep = nBin_p/(max_p-min_p);
	// disegna il contenuto dei bin
	for (int i = 0; i < nBin_p; ++i) {
		std::cout << std::fixed << std::setw(8) << std::setprecision(2) << min_p + i / invStep <<"|";
		int freq = int(scale * binContent_p[i] / max);
		for (int j = 0; j < freq; ++j) {
			std::cout << "#";
		}

		std::cout << std::endl;
	}
	std::cout << "        |\n" << std::endl;
}
